### PR TITLE
UCT/IB/RDMACM: del iface based connection establishment

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
delete iface based connection establishment

## Why ?
out-of-date, replaced with CM based connection establishment

### CODE REVIEW

File: src/uct/base/uct_component.c [hunk 1]
```diff
@@ -105,6 +105,12 @@ ucs_status_t uct_config_read(uct_config_bundle_t **bundle,
     uct_config_bundle_t *config_bundle;
     ucs_status_t status;
 
+    if ((config_table == NULL) || (config_size == 0)) {
```
Comment (alinask) [inline src/uct/base/uct_component.c:108]: why is this needed in this PR?

File: src/uct/base/uct_md.c [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /**
-* Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
+* Copyright (C) Mellanox Technologies Ltd. 2001-2021.  ALL RIGHTS RESERVED.
```
Comment (alinask) [inline src/uct/base/uct_md.c:2]: not sure if it's needed since it's not a new file. @yosefe ?

File: src/uct/base/uct_component.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/uct/base/uct_component.c:108]: gtest segfaults without this change

Comment (alinask) [reply src/uct/base/uct_component.c:108]: yes but why? when is the config_table null?

File: src/uct/base/uct_md.c (hunk 2 above)
Comment (yosefe) [reply src/uct/base/uct_md.c:2]: not mandatory but it's ok to add

File: src/uct/base/uct_component.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/uct/base/uct_component.c:108]: iface config table is NULL now for rdmacm component

**Review by alinask — APPROVED** [review]

**Comment by evgeny-leksikov** [comment]
@yosefe could you pls review

File: src/uct/ib/rdmacm/Makefile.am [hunk 3]
```diff
@@ -15,16 +15,10 @@ libuct_rdmacm_la_LIBADD   = $(RDMACM_LIBS) $(top_builddir)/src/ucs/libucs.la \
                             $(top_builddir)/src/uct/ib/libuct_ib.la
 libuct_rdmacm_la_LDFLAGS  = $(IBVERBS_LDFLAGS) $(RDMACM_LDFLAGS) -version-info $(SOVERSION)
 
-noinst_HEADERS = \
-	rdmacm_md.h \
-	rdmacm_iface.h \
-	rdmacm_ep.h \
-	rdmacm_def.h
+noinst_HEADERS =
```
Comment (yosefe) [inline src/uct/ib/rdmacm/Makefile.am:18]: can remove

File: src/uct/ib/rdmacm/Makefile.am [hunk 4]
```diff
@@ -15,16 +15,10 @@ libuct_rdmacm_la_LIBADD   = $(RDMACM_LIBS) $(top_builddir)/src/ucs/libucs.la \
...
 
 libuct_rdmacm_la_SOURCES = \
-	rdmacm_md.c \
-	rdmacm_iface.c \
-	rdmacm_ep.c
+	rdmacm_component.c
```
Comment (yosefe) [inline src/uct/ib/rdmacm/Makefile.am:21]: the whole rdma_cm component should be compiled-in only if have HAVE_RDMACM_QP_LESS

File: src/uct/ib/rdmacm/rdmacm_component.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "rdmacm_cm.h"
+
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_component.c:12]: another space line

File: src/uct/ib/rdmacm/rdmacm_component.c [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+static ucs_config_field_t uct_rdmacm_cm_config_table[] = {
+    {"", "", NULL,
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_component.c:14]: CM_

File: test/gtest/uct/uct_test.cc [hunk 7]
```diff
@@ -147,46 +147,51 @@ std::vector<uct_test_base::md_resource> uct_test_base::enum_md_resources() {
...
+    component_attr.field_mask = UCT_COMPONENT_ATTR_FIELD_NAME |
+                                UCT_COMPONENT_ATTR_FIELD_FLAGS;
+    /* coverity[var_deref_model] */
+    status = uct_component_query(GetParam()->component, &component_attr);
+    ASSERT_UCS_OK(status);
+
+    UCS_TEST_MESSAGE << "Testing component: " << component_attr.name;
+
+    if (component_attr.flags & UCT_COMPONENT_FLAG_CM) {
+        status         = uct_cm_config_read(GetParam()->component, NULL, NULL,
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc:166]: indent

File: src/uct/ib/rdmacm/rdmacm_component.c [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    return UCS_OK;
+}
+
+uct_component_t uct_rdmacm_component = {
+    .query_md_resources = uct_rdmacm_query_md_resources,
+    .md_open            = ucs_empty_function_return_unsupported,
+#if HAVE_RDMACM_QP_LESS
+    .cm_open            = UCS_CLASS_NEW_FUNC_NAME(uct_rdmacm_cm_t),
+#else
+    .cm_open            = ucs_empty_function_return_unsupported,
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_component.c:45]: no HAVE_RDMACM_QP_LESS -> no component
and can remove all places we check for "unsupported"

File: src/uct/base/uct_component.c (hunk 1 above)
Comment (yosefe) [reply src/uct/base/uct_component.c:108]: why is there a flow which creates iface/md on rdmacm component?
it should not have any iface/md resources

File: src/uct/ib/rdmacm/rdmacm_component.c (hunk 6 above)
Comment (alinask) [reply src/uct/ib/rdmacm/rdmacm_component.c:14]: handled in https://github.com/openucx/ucx/pull/6123, along with tcp_sockcm

**Comment by evgeny-leksikov** [comment]
> why is there a flow which creates iface/md on rdmacm component?
> it should not have any iface/md resources

this is how UCT tests works now, I planned to remove them together with sockcm
anyway I think better do not segfault on reading empty config if we user don't have other way to detect this

File: src/uct/base/uct_component.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/uct/base/uct_component.c:108]: this is how UCT tests works now, I planned to remove them together with sockcm
anyway I think better do not segfault on reading empty config if we user don't have other way to detect this

**Comment by evgeny-leksikov** [comment]
bot:retest

**Comment by evgeny-leksikov** [comment]
bot:pipe:retest

File: src/uct/ib/rdmacm/configure.m4 [hunk 9]
```diff
@@ -29,23 +28,19 @@ AS_IF([test "x$with_rdmacm" != xno],
...
+                                     rdma_init_qp_attr],
+                                    [uct_modules="${uct_modules}:rdmacm"
+                                     rdmacm_happy="yes"
+                                     AS_IF([test "$ucx_check_rdmacm_dir" != /usr],
+                                           [
+                                           AC_SUBST(RDMACM_CPPFLAGS, ["-I$ucx_check_rdmacm_dir/include"])
+                                           AC_SUBST(RDMACM_LDFLAGS,  ["-L$ucx_check_rdmacm_dir/lib$libsuff"])])
+                                     AC_SUBST(RDMACM_LIBS, [-lrdmacm])],
+                                    [AC_MSG_WARN([RDMACM requested but librdmacm is not found])
+                                     AC_MSG_ERROR([Please install librdmacm and librdmacm-devel or disable rdmacm support])
```
Comment (yosefe) [inline src/uct/ib/rdmacm/configure.m4]: `[RDMACM requested but librdmacm is not found or does not provide rdma_establish() API]`

File: src/uct/base/uct_component.c (hunk 1 above)
Comment (yosefe) [reply src/uct/base/uct_component.c:108]: `config_table == NULL` is considered invalid param for this function..
is it possible to disable/skip/remove the UCT tests which cause this?

Comment (evgeny-leksikov) [reply src/uct/base/uct_component.c:108]: Ok I'll will cherry-pick tests clean up which I prepared for next PR with sockcm

Comment (evgeny-leksikov) [reply src/uct/base/uct_component.c]: tests clean up doesn't help, fixed in MD config read

File: src/uct/base/uct_md.c [hunk 10]
```diff
@@ -263,6 +263,10 @@ ucs_status_t uct_md_config_read(uct_component_h component,
     uct_config_bundle_t *bundle = NULL;
     ucs_status_t status;
 
+    if (component->md_config.table == NULL) {
+        goto out;
+    }
```
Comment (yosefe) [inline src/uct/base/uct_md.c:268]: which test/ucp code is calling md_open ith NULL config table?
also, better :
*config_p = NULL;
goto out;

and move out: label to return UCS_OK

Comment (evgeny-leksikov) [reply src/uct/base/uct_md.c:268]: > which test/ucp code is calling md_open ith NULL config table?
> also, better :
> *config_p = NULL;
> goto out;
> 
> and move out: label to return UCS_OK

https://github.com/openucx/ucx/blob/c1560bd8488b6caeada4af9a05f3ad4531b916de/test/gtest/uct/uct_test.cc#L156

Comment (yosefe) [reply src/uct/base/uct_md.c]: need to change uct test, to check UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT and read md config only if it's >0

Comment (evgeny-leksikov) [reply src/uct/base/uct_md.c]: seems working, do you prefer to remove changes in MD?

Comment (yosefe) [reply src/uct/base/uct_md.c]: yes

File: test/gtest/uct/ib/test_sockaddr.cc [hunk 11]
```diff
@@ -23,11 +23,24 @@ class test_uct_sockaddr : public uct_test {
...
         uct_md_h md;
 
+        component_attr.field_mask = UCT_COMPONENT_ATTR_FIELD_NAME |
+                                    UCT_COMPONENT_ATTR_FIELD_FLAGS;
+        /* coverity[var_deref_model] */
+        status = uct_component_query(GetParam()->component, &component_attr);
+        ASSERT_UCS_OK(status);
+
+        if (!(component_attr.field_mask &
+              UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT) ||
```
Comment (yosefe) [inline test/gtest/uct/ib/test_sockaddr.cc:39]: UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT should be added in line 32

Comment (yosefe) [reply test/gtest/uct/ib/test_sockaddr.cc:40]: need to remove UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT from this if

File: test/gtest/uct/uct_test.cc [hunk 12]
```diff
@@ -147,56 +147,71 @@ std::vector<uct_test_base::md_resource> uct_test_base::enum_md_resources() {
...
+    UCS_TEST_MESSAGE << "Testing component: " << component_attr.name;
+
+    if (component_attr.flags & UCT_COMPONENT_FLAG_CM) {
+        status = uct_cm_config_read(GetParam()->component, NULL, NULL,
+                                    &m_cm_config);
+        ASSERT_UCS_OK(status);
+    }
+
+    if ((component_attr.field_mask &
+         UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT) &&
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc:173]: need to remove UCT_COMPONENT_ATTR_FIELD_MD_RESOURCE_COUNT from this if

**Review by yosefe — APPROVED** [review]

**Comment by evgeny-leksikov** [comment]
squashed

File: src/uct/ib/rdmacm/rdmacm_component.c (hunk 6 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_component.c:15]: @evgeny-leksikov merge issue? should be CM_

**Comment by evgeny-leksikov** [comment]
@yosefe ok to merge?

**Comment by yosefe** [comment]
@evgeny-leksikov WDYT about https://github.com/openucx/ucx/pull/6118#discussion_r560471548?

**Review by yosefe — APPROVED** [review]
